File: PrimaMateria/PrimaMateria/Classes/XTRGeneralInfoViewController.m
//
//  GeneralInfoViewController.m
//  PrimaMateria
//
//  Created by Jerry Porter on 11/21/11.
//  Copyright 2016 xTrensa. All rights reserved.
//

@interface XTRGeneralInfoViewController ()
- (void)assignGeneralInfo;
@end

@implementation XTRGeneralInfoViewController

#pragma mark Private Methods

- (void)assignGeneralInfo {
    NSString *path = (self.element).pathForGeneralInfoDoc;
    NSURL *url = [NSURL fileURLWithPath: path];
    NSURLRequest *request = [NSURLRequest requestWithURL: url];
    [self.webView loadRequest: request];
}

#pragma mark - Misc Methods

- (void)setupUI {
    if (self.element != nil) {
        self.discovererLabel.text = [self.element valueForKey: ELEMENT_DISCOVERER];
        self.discoveryLocationLabel.text = [self.element valueForKey: ELEMENT_DISCOVERY_LOCATION];
        self.discoveryYearLabel.text = [self.element valueForKey: ELEMENT_DISCOVERY_YEAR];
        self.abundanceCrustLabel.text = [NSString stringWithFormat: @"%@", [self.element valueForKey: ELEMENT_ABUNDANCE_CRUST]];
        self.abundanceSeaLabel.text = [NSString stringWithFormat: @"%@", [self.element valueForKey: ELEMENT_ABUNDANCE_SEA]];
        [self assignGeneralInfo];
    }
}

#pragma mark - Action Methods

- (IBAction)showWikipediaEntry: (id) sender {
    [self performSegueWithIdentifier: SHOW_WIKIPEDIA_VIEW_CONTROLLER_FROM_GENERAL_VIEW_CONTROLLER sender: self];
}

- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    XTRWikipediaViewController * destController = segue.destinationViewController;
    destController.elementName = self.element.name;
}


#pragma mark - View Management Methods

- (BOOL)shouldAutorotateToInterfaceOrientation: (UIInterfaceOrientation) interfaceOrientation {
    return UIInterfaceOrientationIsLandscape(interfaceOrientation);
}

#pragma mark - Memory Management Methods

- (void)dealloc {
    self.abundanceCrustLabel = nil;
    self.abundanceSeaLabel = nil;
    self.discovererLabel = nil;
    self.discoveryLocationLabel = nil;
    self.discoveryYearLabel = nil;
    self.webView.delegate = nil;
    self.webView = nil;
}

@end

File: PrimaMateria/PrimaMateria/Classes/XTRGeneralInfoViewController.swift
//
//  XTRGeneralInfoViewController.swift
//  PrimaMateria
//
//  Created by Jerry Porter on 4/15/16.
//  Copyright ©2018 xTrensa. All rights reserved.
//

import PDFKit

class XTRGeneralInfoViewController : XTRSwapableViewController {
    
    @IBOutlet var discovererLabel : UILabel!
    @IBOutlet var discoveryLocationLabel : UILabel!
    @IBOutlet var discoveryYearLabel : UILabel!
    @IBOutlet var abundanceCrustLabel : UILabel!
    @IBOutlet var abundanceSeaLabel : UILabel!
    @IBOutlet var webView : UIWebView!
    
    private var progressHUD : MBProgressHUD?
    private var responseData : NSMutableData?
    private var request : NSMutableURLRequest?
    
    var elementName : String?
    
    // MARK: - Initialization Methods
    
    required init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)!
    }
    
    // MARK: - Internal Methods
    
    func assignGeneralInfo() {
        let aPath = element!.pathForGeneralInfoDoc()
        
        if !aPath.isEmpty {
            webView.loadRequest(URLRequest(url: URL(string: aPath)!))
        }
    }
    
    // MARK: - Miscellaneous Methods
    
    override func setupUI() {
        guard let localElement = element else { return }
        discovererLabel.text = localElement.value(forKeyPath: ELEMENT_DISCOVERER) as? String
        discoveryLocationLabel.text = localElement.value(forKeyPath: ELEMENT_DISCOVERY_LOCATION) as? String
        discoveryYearLabel.text = localElement.value(forKeyPath: ELEMENT_DISCOVERY_YEAR) as? String
        abundanceCrustLabel.text = "\(localElement.value(forKeyPath: ELEMENT_ABUNDANCE_CRUST)!)"
        abundanceSeaLabel.text = "\(localElement.value(forKeyPath: ELEMENT_ABUNDANCE_SEA)!)"
        assignGeneralInfo()
    }
    
    // MARK: - Action Methods
    
    @IBAction func showWikipediaEntry(_ sender: UIButton) {
        performSegue(withIdentifier: SegueName.showWikipediaViewControllerFromGeneralViewController, sender: self)
    }
    
    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
        let destController = segue.destination as! XTRWikipediaViewController
        destController.elementName = element!.name
    }
    
    // MARK: - View Management Methods
    
    override var shouldAutorotate : Bool {
        return false
    }
    
    override var supportedInterfaceOrientations : UIInterfaceOrientationMask {
        return .landscape
    }
    
    // MARK: - Memory Management Methods
    
    deinit {
        abundanceCrustLabel = nil
        abundanceSeaLabel = nil
        discovererLabel = nil
        discoveryLocationLabel = nil
        discoveryYearLabel = nil
        webView.delegate = nil
        webView = nil
    }
    
}
